### DISCUSSION TITLE
Is it possible to avoid creation of temp files?

### DISCUSSION THREAD

**Discussion by Jmuccigr** [ImageMagick/ImageMagick#2570] [Help]
I wonder if there's any way to reduce the number of temp files that IM makes, just as a general principle. I see the various security policies, but I'm using it on my laptop from the command line, so I'm not sure those apply.

I ask because I've recently had a few scripts that worked with a larger-than-usual number of files and the amount of disk space that got used up was pretty large.

**Comment by urban-warrior** [comment]
ImageMagick has a policy that restricts the total amount of disk that the pixel cache can consume or the maximum number of frames any image sequence can consume.  If these policies are set and the limits are reached, ImageMagick exits and your command fails.  See https://imagemagick.org/script/security-policy.php for examples.

Your only other option is to reduce the quantum-depth of the ImageMagick builld.  The HDRI build of ImageMagick consumes 4 bytes per pixel component, whereas a non-HDRI 8-bit build only consumes 1 byte per pixel component.  The trade-off is precision.

**Reply by Jmuccigr** [reply]
Thanks. I now see that policy.xml file in my installation, though I also get this unless I edit it, so I assume it's just on the defaults:

```
> identify -list policy

Path: [built-in]
  Policy: Undefined
    rights: None
```

Is there a place to find recommendations for some of these values for personal systems like mine? It's not quite clear to me what they all mean in specific.

**Comment by snibgo** [comment]
Temp files are typically because you don't have enough RAM to store all the images in memory.

So another possibility may be to re-write the commands. For example, a command might read many input images at the start, and then do processing that stores many intermediate results in memory, then write one image to a file and exit. Such a command might be re-written so that at each stage fewer images are in memory at the same time.

**Comment by Jmuccigr** [comment]
Thanks.

Two things I've noticed on my 16GB-RAM laptop:

1. I do have some scripts that read a whole directory of files via something like `magick *.png` and then do something with those files. Clearly as the number of files gets big, IM uses up a lot of temp files. I don't know where it is that those files get so numerous or big that they require writing out, so I guess one question is whether that's configurable.

1. The other place I see temp files is when doing something like mogrify on a group of files and temp files appear in the same directory named with a closing ~, like `image.png~`. I would have assumed that IM is making an intermediate file with all the changes and then deleting the original and renaming the intermediate. The problem with that interpretation is that the intermediate file looks to be a different size than the final version, so it's clearly some intermediate phase of the process. As an example I just took 100 2.4MB files and mogrify'ed them by increasing their size, changing their colorspace and then reducing their size and depth, so that the final file is 151kb. A 2MB intermediate file appears (not sure if it was 2.4) and it seems pretty clear that IM is handling one file at a time, so there should be plenty of room in RAM to hold all that. Even 5 full-sized intermediate files would only be 100 x 5 x 2MB = 1GB. Why can't that just sit in memory until IM is ready to write it out? Or does the default policy.xml not allow that? (It's not clear to me quite how the various parameters in there work exactly.)

**Comment by snibgo** [comment]
Your policy.xml file probably contains many comments, which are demarked by 
```
<!-- and -->
```
...so it doesn't actually create any limits. For a personal system, that is usually fine. If you do unattended work, eg processing movie frames overnight, you might want to set memory limits so if something goes wrong it doesn't create a denial-of-service problem. See https://imagemagick.org/script/resources.php

A command that starts...
```
magick *.png ...
```
... will read all your png files into memory, which is typically unnecessary. Alternatives are to use mogrify to process one file at a time, or use a shell loop and call magick once per file.

My commands are typically complex, and need "magick", not "magick mogrify". So I develop and test with a single file, then use a shell loop to process all the files.

**Reply by Jmuccigr** [reply]
Thanks. I was wondering whether the commenting in profile.xml meant they were inactive or the defaults. So what are the defaults then? (Particularly curious about the default temp dir. It looks like I have some magick- files in my $TMPDIR, so I guess that's it on a Mac.)

Regarding shell loops, yes, that's what I've been doing to avoid the memory issue, but I wrote a bunch of scripts that glob the whole directory that I'll have to redo. :-( Again with 100 1MB files, that's way less than the free RAM on my machine, so I didn't expect a problem.

**Comment by snibgo** [comment]
policy.xml sets limits on resource usage and types of file that can be read or written, and a few other things. The default is "no limit". You can "magick -list resources" to see what resources are available to IM, taking any limits into account.
```
f:\web\im>%IMG7%magick -list resource

Resource limits:
  Width: 107.374MP
  Height: 107.374MP
  Area: 6.40681GP
  List length: unlimited
  Memory: 2.9834GiB
  Map: 5.9668GiB
  Disk: unlimited
  File: 1536
  Thread: 8
  Throttle: 0
  Time: unlimited
```

> Again with 100 1MB files, ...

The filesize in bytes has no relevance to the bytes needed in memory. Files are typically compressed. In memory, assuming v7 HDRI, each pixel with 3 channels takes 12 bytes in memory. So if your 1MB files have 1000x1000 pixels, each will need 12 MB of memory (plus a small overhead). 100 files would need 1.2 GB. After reading an image, you might do operations such as "-blur" that can't be done in-place, so that would need another 1.2 GB.

**Comment by Jmuccigr** [comment]
Thanks, that's helpful. Some more direction in the help pages for the ignorant like me would be welcome.

I'm happy to boost my memory limit to a few gigs, since I typically have that free and it seems like it would help.

Regarding those ~ files, any help?

**Comment by snibgo** [comment]
By my understanding of mogrify.c, the ~ suffix is for a backup filename. This is used to rename an _input_ file, before writing a new new version of that file. If all is well, the backup file is then deleted. If the write has failed for any reason, the backup is not deleted.

I don't have a a reproducible example of a ~ file remaining after a mogrify. If you have one, please share it.

**Comment by Jmuccigr** [comment]
Ah, that makes sense.

It's not that they stick around, it's that I can see them as the script runs. They normally go away. I was just unclear about why they were being made when I could see the original file as well, but that could simply be a lag in the UI.

PS I did get them to stick around by mucking with the temp file location in profile.xml, I think picking a dir that didn't exist.
